cpp/merkletree/sparse_merkle_tree.cc

```diff
@@ -0,0 +1,216 @@
+#include "cpp/merkletree/sparse_merkle_tree.h"
+
+#include <stddef.h>
+#include <algorithm>
+#include <vector>
+
+#include "merkletree/merkle_tree_math.h"
+#include "util/util.h"
+
+using std::make_pair;
+using std::ostream;
+using std::ostringstream;
+using std::reverse;
+using std::string;
+using std::unique_ptr;
+using std::unordered_map;
+using std::vector;
+
+
+const vector<string>* GetNullHashes(const TreeHasher& hasher) {
+  static unique_ptr<const vector<string>> null_hashes;
+  if (!null_hashes) {
+    vector<string> r{hasher.HashLeaf("")};
+    for (int i(1); i < hasher.DigestSize() * 8; ++i) {
+      r.emplace_back(hasher.HashChildren(r.back(), r.back()));
+    }
+    reverse(r.begin(), r.end());
+    null_hashes.reset(new vector<string>(std::move(r)));
+  }
+  return null_hashes.get();
+}
+
+
+SparseMerkleTree::SparseMerkleTree(SerialHasher* hasher)
+    : serial_hasher_(CHECK_NOTNULL(hasher)->Create()),
+      treehasher_(hasher),
+      null_hashes_(GetNullHashes(treehasher_)),
+      root_hash_(null_hashes_->at(0)) {
+}
+
+
+void SparseMerkleTree::EnsureHaveLevel(size_t level) {
+  if (tree_.size() < (level + 1)) {
+    tree_.resize(level + 1);
+  }
+}
+
+
+void SparseMerkleTree::SetLeafByHash(const string& key, const string& data) {
+  serial_hasher_->Reset();
+  serial_hasher_->Update(key);
+  return SetLeaf(PathFromBytes(serial_hasher_->Final()), data);
+}
+
+
+void SparseMerkleTree::SetLeaf(const Path& path, const string& data) {
+  CHECK_EQ(treehasher_.DigestSize(), path.size());
+  // Mark the tree dirty:
+  root_hash_.clear();
+
+  IndexType node_index(0);
+  for (int depth(0); depth <= kDigestSizeBits; ++depth) {
+    int side(PathBit(path, depth));
+    node_index += side;
+    EnsureHaveLevel(depth);
+    auto it(tree_[depth].find(node_index));
+    if (it == tree_[depth].end()) {
+      CHECK(tree_[depth]
+                .emplace(make_pair(node_index, TreeNode(path, data)))
+                .second);
+      return;
+    } else if (it->second.type_ == TreeNode::INTERNAL) {
+      // Mark the internal node hash dirty
+      it->second.hash_.reset();
+    } else if (it->second.leaf_->path_ == path) {
+      // replacement
+      CHECK_EQ(TreeNode::LEAF, it->second.type_);
```

Comment:
If you're here, that's implicitly true, because of line 72, right? And you'd also have already crashed at line 75 (because `it->second.leaf_` would likely have been null)?